runtime/doc/lua.txt

```diff
@@ -1738,6 +1738,33 @@ pesc({s})                                                         *vim.pesc()*
     See also: ~
       • https://github.com/rxi/lume
 
+ringbuf({size})                                                *vim.ringbuf()*
+    Create a ring buffer limited to a maximal number of items. Once the buffer
+    is full, adding a new entry overrides the oldest entry.
+>
+
+    local ringbuf = vim.ringbuf(4)
+    ringbuf.add("a")
+    ringbuf.add("b")
+    ringbuf.add("c")
+    ringbuf.add("d")
+    ringbuf.add("e")      -- overrides "a"
+    print(ringbuf.poll()) -- returns "b"
+    print(ringbuf.poll()) -- returns "c"
```

Comment:
Personally I would go the class based approach here, mostly because it's more conventional Lua and makes it less opaque which is good for inspection/debugging.

```lua
 local ringbuf = vim.ringbuf(4)
 ringbuf:add("a")
 ringbuf:add("b")
 ringbuf:add("c")
 ringbuf:add("d")
 ringbuf:add("e")      -- overrides "a"
 print(ringbuf:poll()) -- returns "b"
 print(ringbuf:poll()) -- returns "c"
```

```lua
function ringbuf:add(item)
    self._items[self._idx_write] = item
    self._idx_write = (self._idx_write + 1) % self._size
    if self._idx_write == self._idx_read then
      self._idx_read = (self._idx_read + 1) % self._size
    end
  end
```

Reply:
My impression is that we are using both depending on the context. Both libuv and ts bindings use the class style. Other ts classes use it too (langtree and highlighter). There are pro's and con's to each.

Reply:
Generics in lua ls don't work properly atm. There's a big tracking issue for them.

Reply:
Changed it to class style.